text/router-enhancements.md

```diff
@@ -0,0 +1,175 @@
+handling constrains manually in the routes requires the app to load and boot the engine in order to
+determine something that could be done much before and save all this expensive work.
+
+Backend fragments like Ruby on Rails get a DRY solution to this allowing to receive an executable
+piece of code that the dynamic segment has to validate in order of that route to match.
+
+In the previous example the constraint would be that the scale is one of our white-listed values, and
+any other would cause the route to miss.
+
+
+# Detailed design
+
+
+### Optional segments.
+
+Mimicking rails micro syntax, I propose optional segments to be enclosed between parenthesis. The
+example of the Apple store would be encoded
+
+```
+Router.map(function() {
+  this.route('product', { path: '(/:region)/:productName' });
+});
+
+
+router.serialize('/mac'); // => routeName='products', params: { productName: 'mac' }
+router.serialize('/uk/mac'); // => routeName='products', params: { productName: 'mac', region: 'uk' }
+```
+
+
+### Dynamic segment's constrains
+
+Also stealing Ruby on Rails' syntax, the user can provide a constraint using the
+config object passed as second argument.
+
+
+```js
+const allowedScales = ['celsius', 'fahrenheit', 'kelvin'];
+
+function isAllowedScale(scaleName) {
+  return allowedScales.includes(scaleName);
+}
+
+Router.map(function() {
+  this.route('forecast', { path: 'forecast/:scale', constraints: { scale: isAllowedScale } });
+});
+
+
+router.serialize('/forecast/kelvin'); // => routeName='forecast', params: { scale: 'kelvin' }
+router.serialize('/forecast/newton'); // => not recognized route.
+```
+
+Although the generic case of receiving a function that returns a boolean is the most generic one,
+it might worth to add a RegExp support as a convenience.
+
+```js
+Router.map(function() {
+  this.route('forecast', { path: 'forecast/:scale', constraints: { scale: /(celsius|fahrenheit|kelvin)/ } });
+
+  # Equivalent to:
+  # this.route('forecast', {
+  #   path: 'forecast/:scale',
+  #   constraints: {
+  #     scale: function(scale) {
+  #       return /(celsius|fahrenheit|kelvin)/.test(scale);
+  #     }
+  #   }
+  # });
+});
+```
+
+# How We Teach This
+
+Expand the guides on the routing topic to cover this use cases.
+
+# Drawbacks
+
+It adds some extra complexity to the routes for a feature that we've survived without for years.
+
+Particularly the dynamic segment constraints, without considering the case of lazy-loaded engines,
+can be emulated without too much effort by manually by checking the validity of the params in the
+routes, and adds some vector of complexity since it's not clear which approach to use as they
+admittedly overlap.
+
+# Alternatives
+
+What other designs have been considered? What is the impact of not doing this?
+
+# Unresolved questions
+
+- Do optional segments have a default value?
+- What arguments should the constraint function receive apart from the dynamic segment itself?
```

Comment:
Things it could receive: the node in the trie representing the route, the handlers, the segment value, and the full set of parsed params. Since I want this to be primarily used for simplistic reduction I'd propose nothing more than the segment value itself.